TOMEE-2891 Translate to Portuguese: examples/resources-jmx-example (#700)
apache/tomee: examples/resources-jmx-example/README_pt.adoc

:index-group: JMX
:jbake-type: page
:jbake-status: status=published
= Custom resources in an EAR archive

O TomEE permite que você defina seus próprios recursos na sua aplicação e declare-os em `META-INF/resources.xml`. Isso permite que você injete esses recursos em qualquer componente gerenciado na sua aplicação.

Além disso, você também pode definir um método `create` no próprio recurso ou em um POJO que atua como uma fábrica.
Este exemplo demonstra o uso do método `create` para registrar adicionalmente o recurso como um JMX MBean, além de disponibilizá-lo para injeção.

== Resource

> Recursos customizados podem ser definidos usando classes Java muito simples. Nesse caso específico, como o aplicativo também deseja registrar esse como MBean, a classe de recurso precisa seguir o MBean especificação.

Reviewer: ```suggestion
Recursos customizados podem ser definidos usando classes Java muito simples. Nesse caso específico, como o aplicativo também deseja registrar esse como MBean, a classe de recurso precisa seguir a especificação de MBean.
```

[source,java]
----
public class Hello implements HelloMBean {

    private AtomicInteger count = new AtomicInteger(0);

    @Override
    public String greet(String name) {
        if (name == null) {
            throw new NullPointerException("Name cannot be null");
        }

        return "Hello, " + name;
    }

    @Override
    public int getCount() {
        return count.get();
    }

    @Override
    public void setCount(int value) {
        count.set(value);
    }

    @Override
    public void increment() {
        count.incrementAndGet();
    }
}

public interface HelloMBean {

    public String greet(final String name);

    public int getCount();

    public void setCount(int count);

    public void increment();

}
----

== Método Criar(Create)

To avoid adding the logic to register the resource as an MBean in every
resource, the application provides a single class with a create() method

> that takes care of this logic for us.

Reviewer: You forgot to remove this lines.

        } catch (final InstanceAlreadyExistsException e) {
            LOGGER.severe("Instance already exists: " + name);
            throw new MBeanRegistrationException(e);
        } catch (final NotCompliantMBeanException e) {
            LOGGER.severe("Class is not a valid MBean: " + code);
            throw new MBeanRegistrationException(e);
        } catch (final javax.management.MBeanRegistrationException e) {
            LOGGER.severe("Error registering " + name + ", " + code);
            throw new MBeanRegistrationException(e);
        }
    }

    private void requireNotNull(final String object) throws MBeanRegistrationException {
        if (object == null) {
            throw new MBeanRegistrationException("code property not specified, stopping");
        }
    }

    public Properties getProperties() {
        return properties;
    }

    public void setProperties(final Properties properties) {
        this.properties = properties;
    }
}
----

NOTE: que esta classe usa as propriedades definidas na configuração
(abaixo), combinada com reflexão, para instanciar o recurso e definir
seus atributos. O código acima requer duas propriedades `code` e `name`

> para saber qual classe criar e o nome JMX para registrá-la
> debaixo.

Reviewer: ```suggestion
para saber qual classe criar e o nome JMX sob o qual será registrada.
```

== Resource

O recurso pode ser definido em `META-INF/resources.xml` da seguinte maneira:

[source,xml]
----
<Resources>
  <Resource id="Hello" class-name="org.superbiz.resource.jmx.factory.JMXBeanCreator" factory-name="create">
    code org.superbiz.resource.jmx.resources.Hello
    name superbiz.test:name=Hello
    count 12345
  </Resource>
</Resources>
----

NOTE: que o atributo class-name se refere à classe factory, e não

> o recurso. Depois que o recurso for criado e vinculado ao TomEE,
> Árvore JNDI da fábrica não é mais usada.

Reviewer: ```suggestion
ao recurso. Depois que o recurso for criado e vinculado à árvore JNDI do TomEE, a fábrica não é mais usada.
```

            return helloMBean.greet(name);
        }

        public void setCount(int count) {
            helloMBean.setCount(count);
        }

        public void increment() {
            helloMBean.increment();
        }

        public int getCount() {
            return helloMBean.getCount();
        }
    }
}
----

O atributo name `<name>/<resource-id>`é usado na anotação `@Resource` para executar a injeção. Nenhuma configuração adicional é necessária para injetar o recurso.

== Propriedades adicionais

Além das propriedades `code` e `name` que o código acima usa para instanciar o recurso, o próprio TomEE fornece algumas propriedades para fornecer mais controle sobre a criação de recursos.

Os recursos são normalmente descobertos, criados e vinculados ao
JNDI muito cedo no processo de implementação, pois outros componentes dependem deles. Isso pode levar a problemas nos quais o classpath final da aplicação ainda não foi determinado e, portanto, o TomEE não pode carregar seu recurso personalizado.

As seguintes propriedades podem ser usadas para alterar esse comportamento.

* Lazy

é um valor booleano que, quando verdadeiro, cria um proxy que adia

> a instanciação real do recurso até a primeira vez que ele olhou para cima do JNDI. 

Reviewer: ```suggestion
a instanciação real do recurso até a primeira vez que ele é procurado pelo JNDI. 
```

Isso pode ser útil se o recurso exigir o classpath da aplicação ou para melhorar o tempo de inicialização não totalmente inicializando recursos que podem não ser usados.

* UseAppClassLoader

Esse valor booleano força um recurso instanciado Lazy a usar o

> classloaders da aplicação, em vez do classloader disponível quando o
> os recursos foram processados primeiro.

Reviewer: ```suggestion
classloaders da aplicação, em vez do classloader disponível quando
os recursos foram processados primeiro.
```

* InitializeAfterDeployment

> Essa configuração booleana força um recurso criado com a propriedade Lazy a ser instanciado depois que a aplicação é iniciado, em vez de esperar que ele seja procurado. Use esse sinalizador se precisar que o recurso seja carregado,independentemente de ser injetado em um componente gerenciado ou consultado manualmente.

Reviewer: ```suggestion
Essa configuração booleana força um recurso criado com a propriedade Lazy a ser instanciado depois que a aplicação é iniciada, em vez de esperar que ele seja procurado. Use esse sinalizador se precisar que o recurso seja carregado, independentemente de ser injetado em um componente gerenciado ou consultado manualmente.
```

Por padrão, todas essas configurações são `false`, a menos que o TomEE encontre um recurso da aplicação personalizado que não possa ser instanciado até que a aplicação seja iniciada. 
Nesse caso, ele definirá esses três sinalizadores para `true`, a menos que o sinalizador `Lazy` tenha sido definido explicitamente.

== PostConstruct / PreDestroy

Como alternativa ao uso de um método de fábrica, você pode usar `@PostConstruct`
e `@PreDestroy` dentro da sua classe de recurso (observe que você não pode
usar isso dentro de uma classe de fábrica) para gerenciar qualquer criação ou
atividades de limpeza. O TomEE chama automaticamente esses métodos quando o
aplicativo é iniciado e destruído. O uso do `@PostConstruct`

> efetivamente forçar um recurso carregado lentamente a ser instanciado quando o

Reviewer: ```suggestion
efetivamente força um recurso carregado lentamente a ser instanciado quando o
```

> aplicativo está sendo iniciado - da mesma maneira que o a propriedade `InitializeAfterDeployment` faz.

Reviewer: ```suggestion
aplicativo está sendo iniciado - da mesma maneira que a propriedade `InitializeAfterDeployment` faz.
```

INFO: A valid shutdown command was received via the shutdown port. Stopping the Server instance.
Apr 15, 2015 12:40:27 PM org.apache.coyote.AbstractProtocol pause
INFO: Pausing ProtocolHandler ["http-bio-61309"]
Apr 15, 2015 12:40:27 PM org.apache.coyote.AbstractProtocol pause
INFO: Pausing ProtocolHandler ["ajp-bio-8009"]
Apr 15, 2015 12:40:27 PM org.apache.catalina.core.StandardService stopInternal
INFO: Stopping service Catalina
Apr 15, 2015 12:40:27 PM org.apache.coyote.AbstractProtocol stop
INFO: Stopping ProtocolHandler ["http-bio-61309"]
Apr 15, 2015 12:40:27 PM org.apache.coyote.AbstractProtocol stop
INFO: Stopping ProtocolHandler ["ajp-bio-8009"]
Apr 15, 2015 12:40:27 PM org.apache.openejb.server.SimpleServiceManager stop
INFO: Stopping server services
Apr 15, 2015 12:40:27 PM org.apache.openejb.assembler.classic.Assembler destroyApplication
INFO: Undeploying app: openejb
Apr 15, 2015 12:40:27 PM org.apache.coyote.AbstractProtocol destroy
INFO: Destroying ProtocolHandler ["http-bio-61309"]
Apr 15, 2015 12:40:27 PM org.apache.coyote.AbstractProtocol destroy
INFO: Destroying ProtocolHandler ["ajp-bio-8009"]

Results :

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
----

> NOTE:as seguintes linhas mostrando a criação do recurso.

Reviewer: ```suggestion
NOTE: as seguintes linhas mostrando a criação do recurso.
```
